src/pq/param.cr

```diff
@@ -1,164 +1,290 @@
+        0
+      )
+
+      slice[0] = val.closed? ? 1u8 : 0u8
+
+      encode_points(val, into: slice + 1)
+      binary slice
     end
 
     def self.encode(val : PG::Geo::Polygon)
-      encode_points "(", val.points, ")"
+      slice = Bytes.new(
+        sizeof(UInt32) +                           # Size
+        sizeof(PG::Geo::Point) * val.points.size + # point data
+        0
+      )
+
+      encode_points val, into: slice
     end
 
-    private def self.encode_points(left, points, right)
-      string = String.build do |io|
-        io << left
-        points.each_with_index do |point, i|
-          io << "," if i > 0
-          io << "(" << point.x << "," << point.y << ")"
-        end
-        io << right
+    private def self.encode_points(val, into slice : Bytes)
+      encode val.points.size, into: slice
+      data = slice + sizeof(UInt32)
+      val.points.each_with_index do |point, index|
+        encode point, into: data + index * sizeof(PG::Geo::Point)
       end
 
-      text string
+      binary slice
     end
 
     def self.encode(val : PG::Interval)
-      # https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT
-      text "#{val.months} months #{val.days} days #{val.microseconds} microseconds"
-    end
+      slice = Bytes.new(sizeof(PG::Interval))
 
-    def self.encode(val)
-      text val.to_s
+      encode val.microseconds, into: slice
+      encode val.days, into: slice + sizeof(Int64)
+      encode val.months, into: slice + sizeof(Int64) + sizeof(Int32)
+
+      binary slice
     end
 
-    def self.binary(slice, size)
-      new slice, size, 1_i16
+    def self.binary(slice, size = slice.bytesize)
+      new slice, size, :binary
     end
 
-    def self.text(string : String)
-      text string.to_slice
+    # Types taken from src/pg/decoder.cr
+    private OID_MAP = {
+      Bool.name                 => 16,   # boolean
+      Bytes.name                => 17,   # bytea
+      Char.name                 => 18,   # char
+      Int16.name                => 21,   # int2
+      Int32.name                => 23,   # int4
+      Int64.name                => 20,   # int8
+      String.name               => 25,   # text
+      Float32.name              => 700,  # float4
+      Float64.name              => 701,  # float8
+      UUID.name                 => 2950, # uuid
+      PG::Geo::Point.name       => 600,  # point
+      PG::Geo::Path.name        => 602,  # path
+      PG::Geo::Polygon.name     => 604,  # polygon
+      PG::Geo::Box.name         => 603,  # box
+      PG::Geo::LineSegment.name => 601,  # lseg
+      PG::Geo::Line.name        => 628,  # line
+      PG::Geo::Circle.name      => 718,  # circle
+      JSON::Any.name            => 3802, # jsonb
+      Time.name                 => 1184, # timestamptz
+      Time::Span.name           => 1186, # interval
+      PG::Interval.name         => 1186, # interval
+    } of String => Int32
```

Comment:
This is what I was hoping to avoid when making decoders accessible from the `PQ` namespace. I used type names as keys rather than the types themselves because I wanted to make it extensible, but I don't know how realistic that is.

I think that, in order to do this extensibly, it would require writing both encoders and decoders, because on top of the OID mapping, it also has to know the size of a given value to encode it, and I ended up making a `SIZE_MAP` below. I don't like that three different data structures have to know about the types available.

This is part of why I don't know whether this PR was a good idea. I may have bitten off more than I had the motivation to chew here and I'm not sure it's worth restructuring the shard to support binary encoding.